grid-proxy/internal/gpuindexer/gpuindexer.go

```diff
@@ -0,0 +1,205 @@
+	}
+
+	sessionId := fmt.Sprintf("tfgrid_proxy_indexer-%d", os.Getpid())
+	client, err := direct.NewClient(ctx, direct.KeyTypeSr25519, mnemonics, relayURL, sessionId, sub, true, indexer.relayCallback)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create direct RMB client: %w", err)
+	}
+	indexer.relayClient = client
+
+	return indexer, nil
+}
+
+func (n *NodeGPUIndexer) queryGridNodes(ctx context.Context) {
+	ticker := time.NewTicker(n.checkInterval)
+	n.runQueryGridNodes(ctx)
+	for {
+		select {
+		case <-ticker.C:
+			n.runQueryGridNodes(ctx)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (n *NodeGPUIndexer) runQueryGridNodes(ctx context.Context) {
+	status := "up"
+	filter := types.NodeFilter{
+		Status: &status,
+	}
+
+	limit := types.Limit{
+		Size:     100,
+		RetCount: true,
+		Page:     1,
+	}
+
+	hasNext := true
+	for hasNext {
+		nodes, _, err := n.db.GetNodes(filter, limit)
+		if err != nil {
+			log.Error().Err(err).Msg("unable to query nodes in GPU indexer")
+			return
+		}
+		if len(nodes) < int(limit.Size) {
+			hasNext = false
+		}
+
+		for _, node := range nodes {
+			id := uuid.NewString()
+			err = n.relayClient.Call(ctx, id, uint32(node.TwinID), "zos.gpu.list", nil)
+			if err != nil {
+				log.Error().Err(err).Msgf("failed to send get GPU info request from relay in GPU indexer for node %d", node.NodeID)
+			}
+		}
+
+		limit.Page++
+	}
+}
+
+func (n *NodeGPUIndexer) gpuBatchesDBUpserter(ctx context.Context) {
+	for {
+		select {
+		case gpuBtach := <-n.nodesGPUBuBatchesChan:
+			err := n.db.UpsertNodesGPU(gpuBtach)
+			if err != nil {
+				log.Error().Err(err).Msg("failed to update GPU info in GPU indexer")
+				continue
+			}
+		case <-ctx.Done():
+			log.Error().Err(ctx.Err()).Msg("Nodes GPU DB Upserter exited")
+			return
+		}
+	}
+}
+
+func (n *NodeGPUIndexer) gpuNodeResultsBatcher(ctx context.Context) {
+	nodesGPUBuffer := make([]types.NodeGPU, 0, n.batchSize)
+	ticker := time.NewTicker(resultsBatcherCleanupInterval)
+	for {
+		select {
+		case nodesGPU := <-n.nodesGPUResultsChan:
+			nodesGPUBuffer = append(nodesGPUBuffer, nodesGPU...)
+			if len(nodesGPUBuffer) >= n.batchSize {
+				log.Debug().Msg("flushing gpu indexer buffer")
+				n.nodesGPUBuBatchesChan <- nodesGPUBuffer
+				nodesGPUBuffer = nil
+			}
+		// This case covers flushing data when the limit for the batch wasn't met
+		case <-ticker.C:
+			if len(nodesGPUBuffer) != 0 {
+				log.Debug().Msg("cleaning up gpu indexer buffer")
+				n.nodesGPUBuBatchesChan <- nodesGPUBuffer
+				nodesGPUBuffer = nil
+			}
+		case <-ctx.Done():
+			log.Error().Err(ctx.Err()).Msg("Node GPU results batcher exited")
+			return
+		}
+	}
+}
+
+func (n *NodeGPUIndexer) Start(ctx context.Context) {
+	for i := 0; i < n.nodesGPUResultsWorkers; i++ {
+		go n.gpuNodeResultsBatcher(ctx)
+	}
+
+	for i := 0; i < n.nodesGPUBufferWorkers; i++ {
+		go n.gpuBatchesDBUpserter(ctx)
+	}
+
+	go n.queryGridNodes(ctx)
+	go n.gpuNodeResultsBatcher(ctx)
+	go n.gpuBatchesDBUpserter(ctx)
```

Comment:
calls to these functions are repeated in the for loop and outside, is there a reason for that?